### ISSUE TITLE
File not download in firefox

### ISSUE THREAD

**Issue by mehulkamani** [jimmywarting/StreamSaver.js#233]
ctrl + shift + r (refresh ) then download file at that time file not download but after simple refresh it work

https://jimmywarting.github.io/StreamSaver.js/localhost/247583/archive.zip get 404

**Comment by Rahul-kalsha** [comment]
Browser Hard Refresh: (Failed)
![image](https://user-images.githubusercontent.com/56481473/125074121-7e37a380-e0da-11eb-8f33-c19726ef13ec.png)

Browser Simple Refresh: (Success)
![image](https://user-images.githubusercontent.com/56481473/125074284-b5a65000-e0da-11eb-970d-884c687c879c.png)

**Comment by jimmywarting** [comment]
Aa, have had some similar issue before in a other package, have something to do with [claiming](https://developer.mozilla.org/en-US/docs/Web/API/Clients/claim) the window

**Comment by jimmywarting** [comment]
hmm, wait... it looks like i'm claiming already

https://github.com/jimmywarting/StreamSaver.js/blob/e00b4a7282a8381a6f72325d6ea93ee1171795da/sw.js#L8

**Comment by Rahul-kalsha** [comment]
Yes @jimmywarting it's already there in sw.js but what's the solution for them because still getting the same issue.

**Comment by jimmywarting** [comment]
Previously i have had grate success using this technique: https://stackoverflow.com/a/60133005/1008999

**Comment by mehulkamani** [comment]
Any solution for this issue ??

**Comment by jimmywarting** [comment]
I can try out the solution from SO, need to change mitm.html + sw.js

**Comment by mehulkamani** [comment]
Any solution for this issue ??

**Comment by mehulkamani** [comment]
Hello Any solution for this issue ??

**Comment by jimmywarting** [comment]
Stop spamming!

**Comment by mehulkamani** [comment]
it not a spamming .just i was asking  any solution

**Comment by mehulkamani** [comment]
If you have no solution, let me know

**Comment by jimmywarting** [comment]
I have already provided a solution that have been working for me in the past in this [SO](https://stackoverflow.com/questions/37559415/how-to-make-serviceworker-survive-cache-reset-shiftf5/60133005#60133005) answer. I said i was going to try it out but haven't gotten around to do it yet.

**Comment by mehulkamani** [comment]
i don't think it working

**Comment by jimmywarting** [comment]
Ok, i tested the old claimMe hack. it did only work for chrome. But it wasn't needed as i did it in the service worker anyway in installation.

However there are a other solution... changing https://github.com/jimmywarting/StreamSaver.js/blob/3955638d238cfca132e2a3d7c3427424cda930f5/StreamSaver.js#L22-L24

to:
```js
const downloadStrategy = isSecureContext && !('MozAppearance' in document.documentElement.style)
   ? 'iframe' 
   : 'navigate' 
```

This will cause a navigation instead of making a iframe to trigger the download. 
the affect of it is that it will trigger a unload event. It might not always be what you want...

Another historical problem (if i remember correctly) with a navigation is that it can potentially freeze the javascript runtime and prevent chunks from being transfered to the service worker. and if you open a download url that have not received any bytes yet then it can create a infinity "waiting for response" before the save dialog shows up and then unfreeze the javascript runtime.

if you can test this out and manage to spot any other side affects in other browsers that would be grate.

A other possible solution could be to open the download url in another popup with `window.open` or `<a href="downloadurl" download="filename" target="_blank">` but that isn't as appealing (i do not remember why i ditched this idea - could be that it can trigger a popup to be blocked) opening a link without `target="_blank"` is the same as navigating

**Comment by mehulkamani** [comment]
const downloadStrategy = isSecureContext && !('MozAppearance' in document.documentElement.style)
   ? 'iframe' 
   : 'navigate' 
   
   Yes this code starts the download but after a while an error occurs.
   
![image](https://user-images.githubusercontent.com/87184347/125888980-55a1b541-35af-491c-bbb7-b2f34e985ae7.png)

**Comment by Rahul-kalsha** [comment]
Hey @jimmywarting 
Can you please suggest me any other solution, I am also stuck in the same situation here. I tried above all solutions but not working well. Also, your [other solution](https://github.com/jimmywarting/StreamSaver.js/issues/233#issuecomment-880861165) **window.open**, which has to allow popups on a manual basis. So it is not user friendly. 

I am appreciated your help but also waiting for your right quick solution. :)

**Comment by jimmywarting** [comment]
**window.open** are often allowed on user interaction... so if you trigger the download durning eg a mouse click (even doe you don't have the data ready at hand) but it often requires you to know what the filename is beforehand.

Another solution I often have to recommend is to use the server whenever possible instead of emulating it with service worker. But it requires you to respond with content-disposition attachment header and you can't use ajax to download it...

Another possible solution is to try out the new https://github.com/jimmywarting/native-file-system-adapter i have built also.
you can save data with streams with or without service worker (service worker are merely a enhancement) + that you have grater control over the service worker so you can pre-install the service worker earlier before you even create a writable download stream

**Comment by mehulkamani** [comment]
any solutions.............................................. @jimmywarting

**Comment by jimmywarting** [comment]
if i had one I would provide one

**Comment by Rahul-kalsha** [comment]
Hey @jimmywarting,  I try everything given by you but not getting any conclusion. I hope the above issue you will be fix your upcoming release. :)

**Comment by WebSpider** [comment]
I would also be really interested in a solution that works with firefox, which is no small browser at all :)

**Comment by WebSpider** [comment]
Thanks for that investigation, and will test if `useBlobFallback` does the trick for me!

**Comment by SilverSeva** [comment]
Hi, any solution on this issue?

**Comment by diafygi** [comment]
An update from my [previous comment](https://github.com/jimmywarting/StreamSaver.js/issues/233#issuecomment-1037780204):

> 1. In Firefox private browsing mode (i.e. incognito mode), `navigator.serviceWorker` is current disabled (see [Firefox Bug #1320796](https://bugzilla.mozilla.org/show_bug.cgi?id=1320796)), so StreamSaver.js will fallback to the normal blob saver functionality (i.e. load everything, then show a file save prompt).

This bug is still outstanding, so StreamSaver.js does not work as intended in Firefox Private Mode (it has to load the full download into memory, then create the download prompt).

Additionally, using [`window.showSaveFilePicker()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/showSaveFilePicker#browser_compatibility) from the File System API is not yet supported in Firefox, so that's not an option either.

Current status: Dynamic stream saving without loading everything into memory is impossible on Firefox Private Browsing mode.

> 2. In Firefox normal mode, when the "Delete cookies and site data when Firefox is closed" setting is checked, `navigator.serviceWorker.register(...)` will throw an exception (see [Firefox Bug #1755167](https://bugzilla.mozilla.org/show_bug.cgi?id=1755167)), which causes the `mitm.html` to silently fail (i.e. says it's downloading, but you never see a download prompt).

This bug is now fixed, so StreamSaver.js works as intended (hooray!).

**Comment by uihp** [comment]
Just to remind, there is another situation that can also cause this error in Chromium broswers.
If your site is using ServiceWorker and you're debugging with DevTools, **"Update on reload" needs to be unchecked**.

**Comment by diafygi** [comment]
> This bug is still outstanding, so StreamSaver.js does not work as intended in Firefox Private Mode (it has to load the full download into memory, then create the download prompt).

As an update, Firefox 140 (released today) fixed [Bug #1320796](https://bugzilla.mozilla.org/show_bug.cgi?id=1320796) now supports service workers in incognito mode by default.

So Firefox now fully supports streaming downloads in both regular and incognito modes! Hooray!

**Comment by ash-m** [comment]
This is not so much a solution as a work-around but just to let everyone know what I did:
```js
const registration = await navigator.serviceWorker.getRegistration();
if (registration && !navigator.serviceWorker.controller) {
    notifications.push({ message: 'A browser error occurred. Please refresh the page normally (without using "shift")', type: 'error' });
    // or soft-refresh automatically
    location.reload();
}
```

Even after a hard refresh, a soft refresh will resync the worker (ime).
